agg-ast/schema_derivation/src/schema_derivation_tests/stage.rs

```diff
@@ -1141,3 +1141,243 @@ mod union {
         })
     );
 }
+
+mod replace {
+    use super::*;
+
+    test_derive_stage_schema!(
+        replace_root_nested_doc,
+        expected = Ok(Schema::Document(Document {
+            keys: map! {
+                "a".to_string() => Schema::Atomic(Atomic::Integer),
+                "b".to_string() => Schema::Atomic(Atomic::String)
+            },
+            required: set!("b".to_string()),
+            ..Default::default()
+        })),
+        input = r#"{"$replaceRoot": {"newRoot": "$name"}}"#,
+        starting_schema = Schema::Document(Document {
+            keys: map! {
+                "name".to_string() => Schema::Document(Document {
+                    keys: map! {
+                        "a".to_string() => Schema::Atomic(Atomic::Integer),
+                        "b".to_string() => Schema::Atomic(Atomic::String)
+                    },
+                    required: set!("b".to_string()),
+                    ..Default::default()
+                }),
+                "foo".to_string() => Schema::Atomic(Atomic::Decimal),
+            },
+            required: set!("foo".to_string(), "name".to_string()),
+            ..Default::default()
+        })
+    );
+
+    test_derive_stage_schema!(
+        replace_root_expression,
+        expected = Ok(Schema::Document(Document {
+            keys: map! {
+                "a".to_string() => Schema::Atomic(Atomic::Integer),
+                "b".to_string() => Schema::Atomic(Atomic::String),
+                "first".to_string() => Schema::Atomic(Atomic::String),
+                "last".to_string() => Schema::Atomic(Atomic::String),
+            },
+            required: set!("b".to_string(), "first".to_string(), "last".to_string()),
+            ..Default::default()
+        })),
+        input = r#"{"$replaceRoot": {"newRoot": {"$mergeObjects": [{"first": "", "last": "" }, "$name"]}}}"#,
+        starting_schema = Schema::Document(Document {
+            keys: map! {
+                "name".to_string() => Schema::Document(Document {
+                    keys: map! {
+                        "a".to_string() => Schema::Atomic(Atomic::Integer),
+                        "b".to_string() => Schema::Atomic(Atomic::String)
+                    },
+                    required: set!("b".to_string()),
+                    ..Default::default()
+                }),
+                "foo".to_string() => Schema::Atomic(Atomic::Decimal),
+            },
+            required: set!("foo".to_string(), "name".to_string()),
+            ..Default::default()
+        })
+    );
+
+    test_derive_stage_schema!(
```

Comment:
Can you add a test where replace_with is using a field which does not appear in the source schema to make sure things are working as expected.

Reply:
Would the test below cover what you are looking for already? it creates new fields using a separate operator and tests that those are correctly in the replaced schema

Reply:
if so this would be the same question for replaceRoot, which also has a test where an expression is being used to replaceRoot rather than simply a field ref

Reply:
You are right, it covers the same idea and I found the unit test for field_ref_missing, so that should be alright 👍